### PR TITLE
Multiple LDA improvements and fixes

### PR DESCRIPTION
**Pull request by jingyimei**
Co-author: Nikhil Kak (nkak@pivotal.io)

This PR addresses the following issues:

JIRAs
MADLIB-1160
MADLIB-1201

1. Ensure that the output of lda_train is consistent with the output of lda_get_word_topic_count
2. Add a helper function, which will map each wordid with corresponding topicid that get assigned in output table.
3. Address LDA topicid index inconsistency issue
4. Fix LDA lda_get_topic_desc getting wrong top_k words issue

All the commits are independent of each other and can be reviewed separately which might be easier than reviewing the files.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/339/

**Comment by fmcquillan99** [comment]
Functional test of these 4 commits seem fine to me.  I added comments and examples in:  
MADLIB-1160
MADLIB-1201

Will create a PR for associated user doc changes shortly.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/340/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/341/

File: src/ports/postgres/modules/utilities/text_utilities.sql_in [hunk 1]
```diff
@@ -74,175 +81,231 @@ tasks related to text.
     Flag to indicate if a vocabulary is to be created. If TRUE, an additional
```
Comment (jingyimei) [inline src/ports/postgres/modules/utilities/text_utilities.sql_in]: a vocabulary table

File: src/ports/postgres/modules/utilities/text_utilities.sql_in [hunk 2]
```diff
@@ -74,175 +81,231 @@ tasks related to text.
...
 
--# Compute the frequency of each word in each document
+-# Compute the frequency of each word in each document:
 <pre class="example">
-DROP TABLE IF EXISTS documents_tf;
-SELECT madlib.term_frequency('documents', 'docid', 'words', 'documents_tf');
-SELECT * FROM documents_tf order by docid;
+DROP TABLE IF EXISTS documents_tf, documents_tf_vocabulary;
+SELECT madlib.term_frequency('documents',    -- input table
+                             'docid',        -- document id
```
Comment (jingyimei) [inline src/ports/postgres/modules/utilities/text_utilities.sql_in]: document id column

File: src/ports/postgres/modules/utilities/text_utilities.sql_in [hunk 3]
```diff
@@ -74,175 +81,231 @@ tasks related to text.
...
+actual word:
 <pre class="example">
-DROP TABLE IF EXISTS documents_tf;
-DROP TABLE IF EXISTS documents_tf_vocabulary;
-SELECT madlib.term_frequency('documents', 'docid', 'words', 'documents_tf', TRUE);
--- Output with wordid instead of the actual words
-SELECT * FROM documents_tf order by docid;
+DROP TABLE IF EXISTS documents_tf, documents_tf_vocabulary;
+SELECT madlib.term_frequency('documents',    -- input table
+                             'docid',        -- document id
```
Comment (jingyimei) [inline src/ports/postgres/modules/utilities/text_utilities.sql_in]: document id column

File: src/ports/postgres/modules/utilities/text_utilities.sql_in [hunk 4]
```diff
@@ -74,175 +81,231 @@ tasks related to text.
     Flag to indicate if a vocabulary is to be created. If TRUE, an additional
     output table is created containing the vocabulary of all words, with an id
     assigned to each word. The table is called <em>output_table</em>_vocabulary
-    (suffix added to the <em>output_table</em> name) and contains the
+    (i.e., suffix added to the <em>output_table</em> name) and contains the
     following columns:
-        - \c wordid: An id assignment for each word
-        - \c word: The word/term
+        - \c wordid: An id for each word.
```
Comment (jingyimei) [inline src/ports/postgres/modules/utilities/text_utilities.sql_in]: We can mention it is in alphabetic ordering.

File: src/ports/postgres/modules/lda/lda.sql_in [hunk 5]
```diff
@@ -182,324 +105,789 @@ lda_train( data_table,
...
     <dt>topic_num</dt>
-    <dd>INTEGER. Number of topics.</dd>
+    <dd>INTEGER. Desired number of topics.</dd>
     <dt>iter_num</dt>
-    <dd>INTEGER. Number of iterations (e.g. 60).</dd>
+    <dd>INTEGER. Desired number of iterations.</dd>
     <dt>alpha</dt>
-    <dd>DOUBLE PRECISION. Dirichlet parameter for the per-doc topic multinomial (e.g. 50/topic_num).</dd>
+    <dd>DOUBLE PRECISION. Dirichlet prior for the per-document topic 
+    multinomial (e.g., 50/topic_num is a typical value to start with).</dd>
```
Comment (jingyimei) [inline src/ports/postgres/modules/lda/lda.sql_in]: I found different libraries do use different starting value, e.g. 1/k, 5/k and 0.1. We can mention this value (50/k) is suggested in Griffiths and Steyvers Paper.

File: src/ports/postgres/modules/lda/lda.sql_in [hunk 6]
```diff
@@ -182,324 +105,789 @@ lda_train( data_table,
...
+                indexed the same as the \c words array above.  For example, if the
+                2nd element of the \c counts array is 4, it means that the word
+                in the 2nd element of the \c words array occurs 4 times in the
+                document.</td>
             </tr>
             <tr>
                 <th>topic_count</th>
-                <td>INTEGER[].</td>
+                <td>INTEGER[]. Array of the count of words in the document
+                that correspond to each topic.</td>
```
[the remark this answers is not in the document]
Comment (fmcquillan99) [reply src/ports/postgres/modules/lda/lda.sql_in]: Array indexing is a developer thing not a user thing, but I will add:

"This array is of length \c topic_num."

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/345/

File: src/ports/postgres/modules/lda/lda.py_in [hunk 7]
```diff
@@ -120,14 +120,22 @@ class LDATrainer:
...
-        if self.iter_num % 2 == 0:
-            work_table_final = self.work_table_0
+        ##This function updates 2 tables, one is the model table and
+        # the other one is the output table
 
+        work_table_final = self.work_table_0 if self.iter_num % 2 == 0 \
+        else self.work_table_1
+
+        # Update model table
+        #JIRA:MADLIB-1201, we have to update model table one more time after
```
Comment (iyerr3) [inline src/ports/postgres/modules/lda/lda.py_in]: I suggest not adding the JIRA number here, because it doesn't add more context.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/348/

**Pull request closed without merging**